auth_mellon_util.c

```diff
@@ -1507,83 +1507,474 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
+                        APR_ARRAY_IDX(tokens, index, Token).offset);
+}
+
+/* This function checks if an HTTP PAOS header is valid and
+ * returns any service options which may have been specified.
+ *
+ * A PAOS header is composed of a mandatory PAOS version and service
+ * values. A semicolon separates the version from the service values.
+ *
+ * Service values are delimited by semicolons, and options are
+ * comma-delimited from the service value and each other.
+ *
+ * The PAOS version must be in the form ver="xxx" (note the version
+ * string must be in double quotes).
+ *
+ * The ECP service must be specified, it MAY be followed by optional
+ * comma seperated options, all values must be in double quotes.
+ *
+ * ECP Service
+ *   "urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp"
+ *
+ * Recognized Options:
+ *
+ * Support for channel bindings
+ *  urn:oasis:names:tc:SAML:protocol:ext:channel-binding
+ *
+ * Support for Holder-of-Key subject confirmation
+ *   urn:oasis:names:tc:SAML:2.0:cm:holder-of-key
+ *
+ * Request for signed SAML request
+ *   urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp:2.0:WantAuthnRequestsSigned
+ *
+ * Request to delegate credentials to the service provider
+ *   urn:oasis:names:tc:SAML:2.0:conditions:delegation
+ *
+ *
+ * Example PAOS HTTP header::
+ *
+ *   PAOS: ver="urn:liberty:paos:2003-08";
+ *     "urn:oasis:names:tc:SAML:2.0:profiles:SSO:ecp",
+ *     "urn:oasis:names:tc:SAML:protocol:ext:channel-binding",
+ *     "urn:oasis:names:tc:SAML:2.0:cm:holder-of-key"
+ *
+ * Parameters:
+ *  request_rec *r              The request
+ *  const char *header          The PAOS header value
+ *  ECPServiceOptions *options_return
+ *                              Pointer to location to receive options,
+ *                              may be NULL. Bitmask of option flags.
  *
  * Returns:
- *   true if the PAOS header is valid, false otherwise
+ *   true if the PAOS header is valid, false otherwise. If options is non-NULL
+ *   then the set of option flags is returned there.
  *
  */
-#define PAOS_VERSION_TOKEN "ver=\"" LASSO_PAOS_HREF "\""
-#define ECP_SERVICE_TOKEN  "\"" LASSO_ECP_HREF "\""
-bool am_validate_paos_header(request_rec *r, const char *header)
+bool am_parse_paos_header(request_rec *r, const char *header,
+                             ECPServiceOptions *options_return)
 {
     bool result = false;
-    char **semicolon_tokens = NULL;
-    char *token = NULL;
-    guint len;
+    ECPServiceOptions options = 0;
+    apr_array_header_t *tokens;
+    apr_size_t i;
+    char *error;
 
-    if (header == NULL) {
-        ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r,
-                     "invalid PAOS header, NULL");
+    ap_log_rerror(APLOG_MARK, APLOG_DEBUG, 0, r,
+                  "PAOS header: \"%s\"", header);
+
+    tokens = tokenize(r->pool, header, true, &error);
+
+#ifdef DEBUG
+    dump_tokens(r, tokens);
+#endif
+
+    if (error) {
+        ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r, error);
```

Comment:
Hi,

sorry for not picking up on this earlier, but this line triggers a warning:

```
auth_mellon_util.c:1872:9: warning: format not a string literal and no format arguments [-Wformat-security]
```

Looking at it more closely, I think this would be a vulnerability, since `error` can contain the actual header that was parsed.

I guess something like this should be a simple solution:

```
ap_log_rerror(APLOG_MARK, APLOG_ERR, 0, r, "%s", error);
```